costmap_2d/plugins/static_layer.cpp

```diff
@@ -80,34 +80,27 @@ void StaticLayer::onInitialize()
   lethal_threshold_ = std::max(std::min(temp_lethal_threshold, 100), 0);
   unknown_cost_value_ = temp_unknown_cost_value;
 
-  // Only resubscribe if topic has changed
-  if (map_sub_.getTopic() != ros::names::resolve(map_topic))
-  {
-    // we'll subscribe to the latched topic that the map server uses
-    ROS_INFO("Requesting the map...");
-    map_sub_ = g_nh.subscribe(map_topic, 1, &StaticLayer::incomingMap, this);
-    map_received_ = false;
-    has_updated_data_ = false;
-
-    ros::Rate r(10);
-    while (!map_received_ && g_nh.ok())
-    {
-      ros::spinOnce();
-      r.sleep();
-    }
-
-    ROS_INFO("Received a %d X %d map at %f m/pix", getSizeInCellsX(), getSizeInCellsY(), getResolution());
 
-    if (subscribe_to_updates_)
-    {
-      ROS_INFO("Subscribing to updates");
-      map_update_sub_ = g_nh.subscribe(map_topic + "_updates", 10, &StaticLayer::incomingUpdate, this);
+  // we'll subscribe to the latched topic that the map server uses
+  ROS_INFO("Requesting the map...");
+  map_sub_ = g_nh.subscribe(map_topic, 1, &StaticLayer::incomingMap, this);
```

Comment:
why does the original function have a check if the topic name has changed? is there a map subscriber elsewhere? i don't see it in this class but perhaps `onInitialize` is called multiple times in the costmap?

Reply:
It's called when costmaps are cleared. With this change the maps are subscribed every time the costmaps are cleared. Depending on how frequently it's called it might not be a desirable thing to do. The reason I removed that check was even after the subscriber is shutdown, the getTopic was returning the topic the subscriber was previously subscribed to (It doesn't check for the state of the subscriber -active or not). I'm thinking to revert back the change and add another check to check the status of subscriber. What do you think?

Reply:
🤔 i know that clearing costmaps is a recovery behavior available in mbf. i would need to look into where it happens in our system and how often, but i think that re-subscribing when they are cleared would be a weird unexpected side effect. i think adding another check is the better course of action.